File: 프로그래머스/Level2/고득점 Kit/소수찾기-완전탐색.py
import math
tmp = []

def dfs(numbers, my_set, visited, cnt, sz) :
    
    if cnt==sz :
        answer = ""
        for t in tmp :
            answer = answer + t
        
        my_set.append(int(answer))
        return
    
    for i in range(0, len(numbers)):
        if visited[i]==1 :
            continue
            
        visited[i] = 1
        tmp.append(numbers[i])
        dfs(numbers, my_set, visited, cnt+1, sz)
        tmp.pop()
        visited[i] = 0
    
    return

def solution(numbers):
    answer = 0
    visited = [0 for _ in range(8)]
    my_set = []
    
    for i in range(0, len(numbers)) :
        dfs(numbers, my_set, visited, 0, i+1)
   
    my_set = set(my_set)
    my_set = list(my_set)
    
    for ele in my_set :
        if ele < 2 :
            continue
        
        check = True
        for i in range(2, ele) :
            if ele%i==0 :
                check = False
                break
        
        if check==True :
            answer += 1
            
    return answer

File: SourceProblem/c++/프로그래머스/Level2/고득점 Kit/소수찾기-완전탐색.cpp
#include <string>
#include <vector>
#include <set>
#include <cmath>
#include <iostream>
using namespace std;

int visited[8] = {0,};
vector<char> v;
set<int> st;
set<int>::iterator it;

void init(){
    for(int i=0; i<8; i++){
        visited[i] = 0;
    }
    v.clear();
    return;
}

void dfs(string numbers, int cnt){
    
    if(v.size()==cnt){
        string tmp = "";
        for(int i=0; i<v.size(); i++){
            tmp = tmp + v[i];
        }
        st.insert(stoi(tmp));
        return;
    }
    
    for(int i=0; i<numbers.size(); i++){
        if(visited[i]==1)
            continue;
        
        visited[i] = 1;
        v.push_back(numbers[i]);
        dfs(numbers, cnt);
        v.pop_back();
        visited[i] = 0;
    }
    return;    
}

int solution(string numbers) {
    int answer = 0;
    int sz = numbers.size();
    
    for(int i=1; i<=sz; i++){    
        dfs(numbers, i);
    }
    
    for(it=st.begin(); it!=st.end(); it++){
        int cur_num = *it;
        
        if(cur_num<2)
            continue;
        
        bool check = true;
        
        for(int i=2; i<=sqrt(cur_num); i++){
            if(cur_num%i==0){
                check = false;
                break;
            }
        }
        
        if(check==true)
            answer++;
    }
    
    return answer;
}